[SourceKit] Move Code Formatting Logic to libIDE

regnerjr opened the pull request:
#### [SourceKit] Move Code Formatting from SourceKit -> libIDE

Moved CodeFormatter, FormatContext, FormatWalker and associated classes, structures, functions to libIDE. 

I have run the tests locally (OS X) `utils/build-script -t`

No new tests have been added since no features were added or removed.

This PR makes progress on - [SR-146](https://bugs.swift.org/browse/SR-146) but does not fulfill all of its scope. 
#### Related bug number: ([SR-146](https://bugs.swift.org/browse/SR-146))

Ping: @akyrtzi @benlangmuir
* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

tkremenek commented:
@swift-ci Please test

tkremenek commented:
@akyrtzi Please review.

benlangmuir commented:
Hi John,

Thanks for working on this!

I think we should squash/rebase some of the commits together to make the changes more independent.  We aim for every commit to be able to build/test successfully on its own.  I don't think we want them all squashed into one commit, but if we can make them more independent/logical it will be easier to review them.

> [SourceKit] Apply failed rebase changes.
> [SourceKit] Add Proper Headers to new files.

These should be rebased into the preceding commits.  Basically the Headers should show up in t he same commit that the new file is added and the rebase change should be squashed into wherever the issue came from.

> [SourceKit] Updated file header to reflect the new file names.
> [SourceKit] Configure Formatting.cpp to be build with libIDE
> [SourceKit] Fix headers to pull in Formatting from libIDE

Similarly, these look like they should be part of preceding commits.  Every commit should independent and be able to build and test successfully on its own.

> [SourceKit] Clean-up header comments.
> [SourceKit] Move FormatContext and FormatWalker to .cpp
> [SourceKit] Formatting - Move More types out of header file.
Should probably be squashed with the commit that moved them in the first place.

> [SoureKit] Limited Formatting Ranges 
Is this new functionality, or is it needed to make the existing tests pass?  If it's new let's put it in a separate pull request (this one is big enough!).  If it's necessary for existing functionality it should probably be squashed together with earlier commits so that they will work on their own.

> [SourceKit] Update Documentation
> [SourceKit] Undo some minor changes

More commits that should be squashed somewhere.

tkremenek commented:
@swift-ci Please test

benlangmuir commented on include/swift/IDE/Formatting.h:
```diff
@@ -0,0 +1,106 @@
+size_t getExpandedIndentForLine(unsigned LineIndex, CodeFormatOptions Options,
+                                StringRef Text);
+
+class SwiftEditorLineRange {
```
How about "SourceLineRange", or even just "LineRange"?  SwiftEditor doesn't really add anything here.

benlangmuir commented on include/swift/IDE/Formatting.h:
```diff
@@ -0,0 +1,106 @@
+/// Request a reformatting of the \p range, using \p Options to determine the
+/// how the format should be applied to \p SF. \p SM is required to provide
+/// an ASTContext and other helper data.
+/// Returns a pair containing which line ranges where updated and a string
```
`\returns`

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+           VD->getParentInitializer()->getEndLoc().isValid() ?
+             VD->getParentInitializer()->getEndLoc() :
+             SourceLoc();
+}
```
Could you run this function through `clang-format`?

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+    bool NeedExtraIndentation;
+
+    class SourceLocIterator : public std::iterator<std::input_iterator_tag,
+    SourceLoc>
```
Please fix the formatting here.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+
+std::pair<SwiftEditorLineRange, std::string>
+reformat(SwiftEditorLineRange Range, CodeFormatOptions Options,
+                                            SourceManager &SM, SourceFile &SF) {
```
Align this with the other parameters.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+
+  auto start = limitedRange.startLine();
+  auto length = limitedRange.lineCount();
+  for( auto currentLine : range(start, start + length) ) {
```
Please format as `for (auto currentLine : range(start, start + length)) {` (note the change in spacing around parens).

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+#include "swift/IDE/Formatting.h"
+
+namespace swift {
+namespace ide {
```
A lot of the code in here isn't actually part of the `swift::ide::` API.  Anything not part of the API should be inside a `namespace {` if it's a type, or marked `static` for global functions.

Also, please switch to `using namespace swift` `using namespace ide` in the implementation file instead of wrapping everything in namespaces.  Then the particular functions you want to expose can be declared qualified like `swift::ide::reformat(...`.  See ModuleInterfacePrinting.cpp for examples.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+}
+
+SwiftEditorLineRange limit(SwiftEditorLineRange Range, SourceManager &SM,
+                                                             unsigned bufferID){
```
Align parameters.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+    auto returned = CF.indent(currentLine, walker, SM, sourceBufferID);
+    output.append(returned);
+  }
+  return std::pair<SwiftEditorLineRange, std::string>(limitedRange, output);
```
Please use `make_pair` instead of the `std::pair` constructor.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -0,0 +1,859 @@
+  for( auto currentLine : range(start, start + length) ) {
+
+    auto returned = CF.indent(currentLine, walker, SM, sourceBufferID);
+    output.append(returned);
```
Please use `raw_string_ostream` instead of repeatedly appending to a `std::string`.

regnerjr commented:
Hi @benlangmuir & @akyrtzi I cleaned up the history and corrected the errors that @benlangmuir pointed out yesterday. Let me know if you see anything that needs more work. Thanks!

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1491,7 +1491,7 @@ class CodeFormatter {
       Builder.append(Line);
       Consumer.recordFormattedText(Builder.str().str());
-      return SwiftEditorLineRange(LineIndex, 1);
+      return std::pair<SwiftEditorLineRange, StringRef> (SwiftEditorLineRange(LineIndex, 1), Text);
```
make_pair

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1530,7 +1530,8 @@ class CodeFormatter {
     // Return affected line range, which can later be more than one line.
-    return SwiftEditorLineRange(LineIndex, 1);
+    SwiftEditorLineRange range = SwiftEditorLineRange(LineIndex, 1);
+    return std::pair<SwiftEditorLineRange, StringRef> (range, Text);
```
make_pair

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1989,8 +1990,8 @@ void SwiftEditorDocument::formatText(unsigned Line, unsigned Length,
-  SwiftEditorLineRange LineRange = CF.indent(Line, FC);
-
+  std::pair<SwiftEditorLineRange, StringRef> indented = CF.indent(Line, FC, Text);
+  SwiftEditorLineRange LineRange = indented.first;
```
SwiftEditorLineRange LineRange = CF.indent(Line, FC, Text).first;

benlangmuir commented on tools/SourceKit/lib/SwiftLang/FormatContext.cpp:
```diff
@@ -43,3 +43,26 @@ size_t swift::ide::getOffsetOfTrimmedLine(unsigned LineIndex, StringRef Text) {
   return LineOffset;
 }
+
+llvm::StringRef swift::ide::getTrimmedTextForLine(unsigned LineIndex, StringRef Text) {
```
You should be able to use `StringRef` without qualifying it here.

benlangmuir commented on tools/SourceKit/lib/SwiftLang/FormatContext.cpp:
```diff
@@ -43,3 +43,26 @@ size_t swift::ide::getOffsetOfTrimmedLine(unsigned LineIndex, StringRef Text) {
+}
+
+size_t swift::ide::getExpandedIndentForLine(unsigned LineIndex,
+                                                     CodeFormatOptions Options,
```
Align parameters.

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1480,7 +1480,7 @@ class CodeFormatter {
     // If having sibling locs to align with, respect siblings.
     if (FC.HasSibling()) {
-      StringRef Line = Doc.getTrimmedTextForLine(LineIndex, Text);
+      StringRef Line = swift::ide::getTrimmedTextForLine(LineIndex, Text);
```
You probably don't need to add `swift::` here.

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -1489,8 +1488,8 @@ class CodeFormatter {
       Builder.append(Line);
-      Consumer.recordFormattedText(Builder.str().str());
-      return std::pair<SwiftEditorLineRange, StringRef> (SwiftEditorLineRange(LineIndex, 1), Text);
+      return std::pair<SwiftEditorLineRange, std::string> (SwiftEditorLineRange(LineIndex, 1),
```
Remove space between > and (.  Also, this should be a call to `make_pair` probably.

benlangmuir commented on include/swift/IDE/Formatting.h:
```diff
@@ -0,0 +1,107 @@
+size_t getExpandedIndentForLine(unsigned LineIndex, CodeFormatOptions Options,
+                                StringRef Text);
+
+class LineRange {
```
This seems to have been renamed in the same commit as moving the code into place.  It would be great if it happened earlier so that this commit was really just moving code around.  If there are other modifications happening here, I think the same applies unless the changes are somehow tied to the act of moving the code.  Does that make sense?

benlangmuir commented on include/swift/IDE/Formatting.h:
```diff
@@ -0,0 +1,107 @@
+#include "swift/Frontend/Frontend.h"
+#include "swift/Basic/SourceManager.h"
+#include "swift/IDE/SourceEntityWalker.h"
+#include "swift/Subsystems.h"
```
It's good practice to minimize includes in headers, possibly using forward declarations where appropriate.  I think most of these could be sunk into the .cpp file.

benlangmuir commented:
This is looking awesome!  I've made a few more suggestions on the individual commits here.  Thanks for your hard work!

regnerjr commented:
Hey @benlangmuir! Thanks so much for you help. I am now a rebase, rewriting history master. Thanks for the great suggestions. I think I have now addressed all your comments, and ensured that each individual commit, builds and passes all tests. Let me know if there is anything more you'd like to see here. Thanks again.

benlangmuir commented:
@swift-ci Please test
